Indirect precode through slot

vitek-karas opened the pull request:
High-level changes:
- All calls within the same NGEN module go through indirection
- Method slots are used as the indirection cells
- Changes the slots to be writable and absolute pointers (both on MD and MT)
- Methods saved into native image are never marked as having a precode. If they don't have one, they are marked as stable entry point (as today). If they do have one, they are mark as not having a stable entry point. This behaves similarly to the method having a temporary entry point.
- Fixes around the fact that such native methods don't have the real temporary entry point - so special case these and use the "old" value of the slot as the temporary entry point (which points to the saved precode).

Passes most CoreCLR tests, some might still fail - work in progress.

Unsolved:
- Special methods (unboxing stubs, com interop) require creation of runtime allocated precodes at places where we don't normally backpatch all possible slots for the method. This means we might leave around code/slots which point to the saved precode which will never be patched. Slow since it would always go through prestub.

Future:
- Detect which methods need precodes at NGEN time and avoid indirection calls to those which don't. For now probably by "hacking" the callsite. Future - some methods can be decided up front, fixup only methods could be solved by branching in JIT.

vitek-karas committed:
Change zapped precodes to modify MD slot when patched instead of modifying the precode itself.

vitek-karas committed:
Mapping of method slots to zap nodes.

Registers each non-virtual method slot with the pointer->node mapping in DataImage.
Introduce new ZapMethodSlot node which is a placeholder for the method slot address.
Use the pointer->node mapping to lookup the RVA for the ZapMethodSlot.

vitek-karas committed:
Allows patching of slots instead of precodes for native images.

Modifies method table and vtable chunks to be writeable.
Methods with precode are saved as methods with "no-repcode" and "no-stable entrypoint". This emulates temporary entry points.
Modified places in MD where we assume that no stable entry point means temporary entry point, which is no longer the case - for zapped ethods this is a valid state and the "temporary entry point" is the value in the MT slot.
Mark all precodes as RX only.

With this the system works while zapped precodes are never modified.

vitek-karas committed:
Make all calls between two methods in the same native module go through indirection cell.

Change calls in the same module for NGENed modules to go through indirection, using the method's slot as the indirection cell.
Fixes the ZapMethodSlot node to disambiguate it from other method based wrappers.
Even virtual vtable slots in MT must be registered as those are also slots potentially used by the new node.
Improve backpatching in the zapped case - if we're patching a zapped method from a zapped precode to something else (the new stable entry point), then go and patch all potential slots for that method, even though it doesn't really have a temporary entry point. Basically treat zapped precodes as temporary entry points for the backpatching logic.

vitek-karas committed:
Two fixes to handle special methods.

Unboxing stub and com interop. The same problem in both cases, the code expects to have a stable entry point with a precode in these cases, but with the RX changes that's not the case.
So solving this by calling GetOrCreatePrecode to force runtime allocated precode.
Needs more work around patching all the possible slots with this.

jkotas commented on src/vm/method.cpp:
```diff
@@ -2149,12 +2146,29 @@ PCODE MethodDesc::TryGetMultiCallableAddrOfCode(CORINFO_ACCESS_FLAGS accessFlags
+            // a zapped precode. That precode is not patchable, so we can't make it a stable entry point.
+            // So instead force creation of runtime allocated precode which is patchable and which will
+            // become the stable entry point.
+            // TODO: How to solve the problem of patching all possible MT slots for this method.
```
`MethodDesc::DoBackpatch` will keep looking for more unpatched slots every time it gets called. We just need to make sure that it works as expected for these.

jkotas commented on src/vm/method.cpp:
```diff
@@ -2149,12 +2146,29 @@ PCODE MethodDesc::TryGetMultiCallableAddrOfCode(CORINFO_ACCESS_FLAGS accessFlags
     // We create stable entrypoints for these upfront
     if (IsWrapperStub() || IsEnCAddedMethod())
+    {
+        if (IsZapped() && !HasStableEntryPoint())
```
This condition should not be needed. We should be able to just call GetOrCreatePrecode() all the time. (It would be tiny bit more efficient.)

jkotas replied:
(It can be just assert.)

jkotas commented on src/vm/method.cpp:
```diff
@@ -4769,14 +4814,32 @@ Precode* MethodDesc::GetOrCreatePrecode()
+        // instead the precode acts as a temporary entry point.
+        pExpected = *pSlot;
+#ifdef FEATURE_PREJIT
+        Module * pZapModule = GetZapModule();
```
I think there is a race if other thread calls GetOrCreatePrecode at the same time - the slot is not going to point to zapped precode anymore.

jkotas commented on src/vm/method.cpp:
```diff
@@ -4848,8 +4911,28 @@ BOOL MethodDesc::SetStableEntryPointInterlocked(PCODE addr)
+        pExpected = *pSlot;
+#ifdef FEATURE_PREJIT
+        Module * pZapModule = GetZapModule();
+        _ASSERTE((pZapModule != NULL) && pZapModule->IsZappedPrecode(pExpected));
```
Same here

jkotas commented on src/vm/precode.h:
```diff
@@ -256,6 +260,12 @@ class Precode {
+    BOOL IsZapped();
+    // Determines if the precode is part of a native image
+    // Optimization if the caller already has the precode's MethodDesc available.
+    BOOL IsZapped(PTR_MethodDesc pMD);
```
I do not think that this optimization is worth it. Fetching methoddesc from precode is cheap.

jkotas commented:
There seems to be way too many places where the `IsZapped` is special cased. I am wondering whether it would be better to really save the throw-away precodes into the NGen image as real temporary entrypoints (ie MethodDescChunk would have a pointer to the block just like in the JITed case, etc.).

vitek-karas committed:
PR feedback fixes

Fix races when swapping zapped precodes with stable entry points.
Simplify code in precodes by removing unnecessary optimizations.
Simplify prestub by using GetOrCreatePrecode instead of more compelx code which will end up doing the same anyway.

vitek-karas commented:
Good news: Like this the changes pass all CoreCLR tests (runtests that is).

vitek-karas committed:
Simplify the code change by not allowing SetTarget on zapped precodeds.

Turns out that zapped precodes can't have the SetTarget called on them with one exception the backpatch. In that case doing it is moot anyway as it would not do anything (we're setting it to the stable entry point which is already in the main method slot at that time).
So I removed the ability to call SetTarget on zapped precodes completely.

jkotas commented on src/vm/method.cpp:
```diff
@@ -4768,15 +4809,40 @@ Precode* MethodDesc::GetOrCreatePrecode()
+        // If the method is zapped, then the only way it may not have a stable entry point yet
+        // is that it has a precode. In this case it's not marked as having precode through HasPrecode
+        // instead the zapped precode acts as a temporary entry point.
+        pExpected = *pSlot;
```
I know we probably do this in other places ... but this pattern is prone to be broken by C/C++ compiler optimizations.

https://github.com/dotnet/coreclr/pull/16089 is a recent example 

The fix is to use VolatileLoadWithoutBarrier to load the value. Like: `pExpected = VolatileLoadWithoutBarrier(pSlot);`

jkotas commented on src/vm/method.hpp:
```diff
@@ -1634,11 +1634,13 @@ class MethodDesc
     //     pDispatchingMT - method table of the object that the method is being dispatched on, can be NULL.
     //     fFullBackPatch - indicates whether to patch all possible slots, including the ones 
     //                      expensive to patch
+    //     pPreviousEntryPoint - if this is not NULL, the value is used as the expected value for all the patching
```
Is `pPreviousEntryPoint` just an optimization, or is it required for correctness?

I am worried that there can be bad (crashing?) races between `Prestub` and `DoBackpatch` because of this.

vitek-karas replied:
My understanding is that this is actually required for correctness.
Basically DoBackpatch needs to determine the expected "old" value of the method entry point. For method with temporary entry point we use that - assuming that the entry point only ever changes from temporary to stable and never again. So we blindly take the temporary and see if something matches it - if so we patch that.
With the zapped methods, we don't have a place to lookup the address of the zapped precode, since by the time we get to DoBackpatch the method slot is already overwritten with the stable entry point. So I had to work around that.
The way it works is that SetStableEntryPointInterlocked returns the previous value (the zapped precode, we actually verify this in the method itself). This previous value is then passed to DoBackpatch. DoBackpatch will then use it as the temporary entry point, that is it will see if there are other slots with the previous value, and if so, it will patch them with the stable entry point.
This actually works for both zapped and non-zapped methods, since the logic in SetStableEntryPointInterlocked must be the same as the one in DoBackpatch.

Regarding races:
- SetStableEntryPointInterlocked will return temporary entry point for non-zapped method in this out-parameter. So the behavior is exactly the same as before (we just don't load it again in DoBackpatch, but take it as a parameter instead).
- SetStableEntryPointInterlocked will return the previous slot value for zapped methods, but only if it passes the IsZappedPrecode condition. So if there's a race and the stable entry point is set twice, the second call will see the new value in the slot, but that will not pass the IsZappedPrecode - in which case we would return NULL as the previous value -> leave DoBackpatch do what it did so far (which for zapped methods is effectively nothing in this case since the method is not marked as having a temporary entry point).
- If it so happens that the second call to SetStableEntryPointInterlocked still sees the old value there - the zapped precode - it will return that. DoBackpatch will use it just like it does temporary entry point, potentially trying to backpatch the method for the second time. But since it does so by reading the stable entry point and then patching with that, at worst it would patch to the same value - in reality, that is unlikely to happen, since the old value will not match what's in the slots anymore.

vitek-karas committed:
Prevent races due to compiler optimizations

jkotas commented:
```csharp
using System;

public class Parent {
    public virtual string foo() {
        return "Hello";
    }
}

public class Child : Parent {
}

public class Test {
    static void Main() {
          new Parent().foo(); // This will patch Parent::foo slot as expected

          // What is going to patch Child::foo slot? If I am reading the code correctly, 
          // this will always go through prestub. We will exit early from DoBackPatch
          // via `if (!hasZappedPrecodeEntryPoint && !HasTemporaryEntryPoint())`
          for (;;) { new Child().foo(); }
    }
}
```

vitek-karas commented:
With a simple tweak (add a virtual to Child to avoid vtable chunk sharing) this does indeed break the patching. I didn't realize that on the first call into DoBackpatch we do have the right values, but that one is the partial backpatch. On the second call (through Child) we don't have the right values and we bail out as you mentioned.
I'm gonna think about this some more, but currently the only viable solution is to really introduce temporary entry points (that is save the zapped precodes into a chunk pointed to by the method desc chunk). The reason is not to share the code with existing temporary entry point infra (might be able to do some of that as well though), but so that we can always ask "What was the original precode used by this method" so that we can always backpatch correctly.... :-( oh well.

vitek-karas committed:
Repurpose ZapMethodEntryPoint as ZapMethodSlot

Instead of introducing new ZapMethodSlot and leaving the ZapMethodEntryPoint unused - rename ZapMethodEntryPoint to ZapMethodSlot - also the cache is reused. This is the actually purpose of this change since the cache uses both method handle and access flags for lookup (as oppose to the wrappers which only use the method handle).
This will be needed once we start replacing call indirections with direct calls for methods which can be called directly (Where the decision relies on the access flags as well).
Some minor cleanup.

vitek-karas committed:
A bit more cleanup after the ZapMethodEntryPoint to Slot change

vitek-karas committed:
Add the abitlity to store temporary entry points in NGEN on MD chunk.

Basically saves teh MD chunk with one additional pointer in front of it. That's where we will store the temporary entry points pointer (just like the runtime allocated MD chunks do).
In order to do this cleanly we need to be able to map the MD chunk pointer to an ZapInnerPtr node (and not the structure node which backs up the memory since that is starting eith the temporary entry point pointer).
This introduces the ZapInnerPtr nodes to lot of the DataImage Fixup and BindPointer infra. So adapt that infra to be able to handle inner ptr nodes correctly.

vitek-karas committed:
Save precodes as temporary entry points

- Change precode saving to use the exact same layout as temporary entry points
- This also means that we save only one type of precode (either fixup or stub depending on defines).
- Store the saved precodes as the value for the temporary entry points slot in the MD chunk

No runtime changes yet, other than the fact that precodes for methods are now comming from a different place. Saved methods are still treated as not having temporary entry points.

vitek-karas committed:
Save temporary entry points for all methods in a chunk

Change chunk method ordering to consider the need for precodes - with this we try to put methods which need precodes together and split chunks on that. So we should have chunks where all (or most) methods need precodes and chunk where none do.
Save precodes only for chunks where at least one method needs them.

Register all precodes as surrogates of the "new" MD, so that we can run fixup on them (easy lookup). Leave the current registration of precodes as surrogates of the "old" MD only for methods which need precodes. This is used to figure out if the method should use precode in its MT slot.

vitek-karas committed:
Clean up precode marking during MD save

vitek-karas committed:
Remove unnecessary special casing for zapped methods

Remove the special code to deal with zapped precodes in GetOrCreatePrecode and SetStableEntryPoint. In both cases we can now rely on the GetTemporaryEntryPoint to work correctly. The only exceptrion is GetOrCreatePrecode which still has to enfore new prcode creation over zapped precode, regardless of the type of the zapped precode.
Also removed the special code in DoBackpatch as it can also rely on GetTemporaryEntryPoint.

vitek-karas committed:
Cleanup - remove unused code for saving precodes

vitek-karas committed:
PR feedback: Switch from using inner ptr to save MethodDescChunk to an approach based on offsets

Introduced some constant so that code outside of the MD chunk saving logic can use those instead of relying on knowledge of the actual layout of the nodes.
Removed the DataImage changes to support inner ptr pointers with the Fixup... methods as those are not needed anymore.

vitek-karas committed:
Methods which need remoting precode will get one in prestub without running the rest of prestub.

vitek-karas committed:
Fix remoting precode creation in Prestub

Only force create remoting precode if the method doesn't have a stable entry point yet. This should really only happen for zapped methods, but it would work on all.
On subsequent calls to prestub for the same method, it would already have a stable entry point and thus the normal perstub logic would execute.

vitek-karas committed:
Merge pull request #5 from vitek-karas/TemporaryEntryPointsInNGEN

Temporary entry points in ngen

vitek-karas committed:
Fixes for building and running AMD64

Couple of ULONG/SIZE_T/DWORD casting issues - also fixed usage of SIZE_T versus SSIZE_T.
Fixed a bug in DoBackpatch which still conatined remnants of reverted change. Make it so that zapped precodes are not attempted to be patched.

vitek-karas commented:
I'm merging this to keep future diffs reasonable. The change passes both X86 and AMD64 CoreCLR tests and also all of my specific tests, including the backpatching test mentioned below (verified in the debugger that we correctly backpatch and don't repeatedly go through prestub).
